Implement the SSE hardware intrinsics.

tannergooding opened the pull request:
This implements basic codegen support for all currently declared SSE intrinsics except for the `Store` intrinsics.

tannergooding commented on src/jit/emitxarch.cpp:
```diff
@@ -107,7 +108,7 @@ bool emitter::IsDstSrcSrcAVXInstruction(instruction ins)
 {
     return IsAVXInstruction(ins) &&
            (ins == INS_movlpd || ins == INS_movlps || ins == INS_movhpd || ins == INS_movhps || ins == INS_movss ||
-            ins == INS_movlhps || ins == INS_sqrtss || ins == INS_sqrtsd || ins == INS_cvtss2sd || ins == INS_cvtsd2ss);
```
`INS_movlhps` was incorrectly listed as `DstSrcSrc`, which will cause it to generate a different result on AVX vs non-AVX machines.

tannergooding commented on src/jit/emitxarch.cpp:
```diff
@@ -4812,6 +4813,22 @@ void emitter::emitIns_SIMD_R_R_R(instruction ins, regNumber reg, regNumber reg1,
+
+void emitter::emitIns_SIMD_R_R_R_I(instruction ins, regNumber reg, regNumber reg1, regNumber reg2, int ival, var_types simdtype)
+{
+    if (UseVEXEncoding() && reg1 != reg)
```
For both this and `emitIns_SIMD_R_R_R` above: Do we have a work item tracking support for `reg2` being a memory operand when it is properly aligned?

Currently we are generating things like
```
C4E1791000           vmovupd  xmm0, xmmword ptr [rax]
C4E14858F0           vaddps   xmm6, xmm0
```

While in some cases we could generate `vaddps xmm6, [rax]`

mikedn replied:
> reg2 being a memory operand when it is properly aligned?

And in what circumstances will the memory operand be aligned? Only for local variables I guess.

tannergooding replied:
Yeah. It definitely can't be used everywhere, and isn't something I think needs to be immediate.

But, for the cases where the memory will always be aligned (locals, is one example), it would be useful to generate the more efficient code

4creators replied:
> Only for local variables I guess

@tannergooding @mikedn
 
It is not true. In general devs writing vectorized code would peel off start of loops to align data and do all the remaining calculations on aligned data until the end of the loops where processing may require change of logic.

This means  that all data accessed in the `main` loops will be aligned and doing anything but `vaddps xmm6, [rax]` would be an unacceptable loss of the CPU cycles.

mikedn replied:
> In general devs writing vectorized code would peel off start of loops to align data

The key here is "in general". The JIT has no way to know that the address is aligned or not and it cannot make such assumptions simply because "in general" developers do that.

I suspect that it will work if the developer uses `LoadAligned` to indicate that the memory address is aligned.

tannergooding replied:
I agree that, "in general", users writing performant code will try to keep accesses aligned to avoid reads/writes that cross cache line or page boundaries.

However, as @mikedn points out, assuming this is incorrect and detecting it is likely hard.

It's probably worth a separate discussion (will log a bug in a bit) but I believe it is something we should try to enable (maybe a special intrinsic, compiler hint,  or attribute to tell the jit memory accesses will be aligned for a given method).

4creators replied:
> It's probably worth a separate discussion (will log a bug in a bit) 

It is the right approach I think as this not that easy to get it right

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -103,6 +103,133 @@ void CodeGen::genSSEIntrinsic(GenTreeHWIntrinsic* node)
             emit->emitIns_SIMD_R_R_R(INS_addps, targetReg, op1Reg, op2Reg, TYP_SIMD16);
             break;
+
+        case NI_SSE_And:
```
All the cases (so far) are essentially the same with only the instruction (or immediate value) differing.

It might be worthwhile including this info in the `hwintrinsiclist` and having them be entries in an array for constant time lookup...

mikedn replied:
Yes, unlike "classic" instructions SSE/AVX instructions have a more consistent format so it should possible to drive most codegen from tables. Basically you need to add to the list a "form" value that can be used to select between `R_R_R` and `R_R_R_I` and of course, the instruction itself.

Well, there's also the problem of memory operands but that's something that the IR node will indicate. So `R_R_R` in the hw intrinsic list really stands for `R_R_RM`.

tannergooding replied:
I was actually just thinking of having a single array, something where the entries are {instruction, immediate}. Helper functions (like `get_IsSupported`) would have `INS_INVALID` and instructions without an immediate would probably have `-1`.

There can then be two general groupings. One for instructions without an immediate (those that use `SIMD_R_R_R`) and one for instructions with an immediate (those that use `SIMD_R_R_R_I`).

Other instructions (such as `Load`) which have different semantics can be in their own case grouping.

I don't think we necessarily need a `form` value, unless you think that would be more efficient.

fiigii replied:
I do not want to tabulate the codegen of hw intrinsic, which won't always be such simple. 
For example, `intrinsicID` + `base` is not enough to indicate codegen for some intrinsic, I would add a new field into `GenTreeHWIntrinsic` to represent overloads that would be used here.

mikedn replied:
> I don't think we necessarily need a form value, unless you think that would be more efficient.

The `form` thing was just a suggestion. Any solution that cuts down the amount of repeated code should be good enough, provided that it doesn't cause other problem.

>  I would add a new field into GenTreeHWIntrinsic to represent overloads that would be used here.

Hmm? It's not clear why that would be better.

tannergooding replied:
I don't think we should do it for all instructions, but a number of them (especially the binary arithmetic and comparison operators) will all be fairly common and follow roughly the same code path.

Cutting down duplicate code will likely be worthwhile for these operations.

fiigii replied:
> Hmm? It's not clear why that would be better.

@mikedn For example, we need an additional field to distinghuish these two intrinsics (`intrinsicID` + `baseType` is not enough).
```csharp
Vector128<float> GatherVector128(float* baseAddress, Vector128<long> index, byte scale);
Vector128<float> GatherVector128(float* baseAddress, Vector128<int> index, byte scale);
```
My current is to add a `overloads` field in `GenTreeHWIntrinsic` to bring that information from importer to CodeGen.

fiigii replied:
Here is the complete list of `GatherVector128` to show the necessity.
```CSharp
        /// <summary>
        /// __m128i _mm_i32gather_epi32 (int const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i32gather_epi32 (int const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<uint> GatherVector128(uint* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i32gather_epi64 (__int64 const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<long> GatherVector128(long* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i32gather_epi64 (__int64 const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<ulong> GatherVector128(ulong* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128 _mm_i32gather_ps (float const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<float> GatherVector128(float* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128d _mm_i32gather_pd (double const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<double> GatherVector128(double* baseAddress, Vector128<int> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i64gather_epi32 (int const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i64gather_epi32 (int const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<uint> GatherVector128(uint* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i64gather_epi64 (__int64 const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<long> GatherVector128(long* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128i _mm_i64gather_epi64 (__int64 const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<ulong> GatherVector128(ulong* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128 _mm_i64gather_ps (float const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<float> GatherVector128(float* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
        /// <summary>
        /// __m128d _mm_i64gather_pd (double const* base_addr, __m128i vindex, const int scale)
        /// </summary>
        public static unsafe Vector128<double> GatherVector128(double* baseAddress, Vector128<long> index, byte scale) => GatherVector128(baseAddress, index, scale);
```

mikedn replied:
> intrinsicID + baseType is not enough

Hmm, can't we use a different intrinsic id for each overload?

fiigii replied:
> Hmm, can't we use a different intrinsic id for each overload?

That will break the intrinsic searching because a method name has to map to one intrinsic id in the current `[Intrinsic]` system.

tannergooding commented on src/jit/hwintrinsiclistxarch.h:
```diff
@@ -11,58 +11,96 @@
+HARDWARE_INTRINSIC(SSE_Multiply,                        "Multiply",                     SSE)
+HARDWARE_INTRINSIC(SSE_Or,                              "Or",                           SSE)
+HARDWARE_INTRINSIC(SSE_Reciprocal,                      "Reciprocal",                   SSE)
+HARDWARE_INTRINSIC(SSE_ReciprocalSqrt,                  "ReciprocalSquareRoot",         SSE)
```
This name is getting fixed in https://github.com/dotnet/coreclr/pull/15471, will update after it is merged.

tannergooding commented:
I split and categorized the hardware intrinsic tests to make it easier to tell what coverage exists for each ISA and each instruction.

A good portion of the code could be shared and it might be worth refactoring sooner rather than later.

tannergooding commented:
For the `CompareGreaterThan` intrinsics, the hardware manual recommends using the inverse operations or by using software emulation (swapping the operands and using a different predicate).

Using the inverse operations is preferred, correct? I was trying to think of reasons why software emulation would be suggested, but came up drawing a blank....

tannergooding commented:
FYI. @fiigii

I figured I would start on this while waiting for the scalar intrinsic PR to get merged. I believe you are going to be focusing on AVX/AVX2, so I hopefully am not duplicating work you've already done here 😄

mikedn commented on src/jit/emitxarch.cpp:
```diff
@@ -4812,6 +4813,23 @@ void emitter::emitIns_SIMD_R_R_R(instruction ins, regNumber reg, regNumber reg1,
+void emitter::emitIns_SIMD_R_R_R_I(
+    instruction ins, regNumber reg, regNumber reg1, regNumber reg2, int ival, var_types simdtype)
+{
+    if (UseVEXEncoding() && reg1 != reg)
```
Parenthesize reg1 != reg

fiigii commented:
@tannergooding Thank you so much for the work! 

>  so I hopefully am not duplicating work you've already done here 

My next step is to implement `Load`/`Store` of AVX/AVX2/SSE/SSE2 because these memory-access intrinsics need a bit more infrastructure.

I recommend you start with simple intrinsics that generate single instruction (e.g., arithmetic, bit manipulation, and comparison, etc.). Then certain helper intrinsics can be expanded to combinations of these simple intrinsics.

fiigii commented:
> For both this and emitIns_SIMD_R_R_R above: Do we have a work item tracking support for reg2 being a memory operand when it is properly aligned?

Yes, this feature is in my plan, we have to enable containment analysis on hardware intrinsic in the future. In C++, users can indicate this codegen by deferring pointers of vectors, but C# has no pointer to managed type. So this feature would considerably impact CQ.
BTW, not only "reg2 being a memory operand", but also reg1 to fold `Store*`.

mikedn commented:
> but C# has no pointer to managed type

Hmm, it does have that - `ref`. Aka `byref` in JIT.

fiigii commented:
> this feature is in my plan, we have to enable containment analysis on hardware intrinsic in the future

To be clear.
```csharp
var res = Avx.Add(vec1, Avx.Load(address));
```
should generate
```asm
vaddps ymm1, ymm2, [adress]
```

mikedn commented:
> To be clear...

Yes. Though I suppose it should also work with `Avx.Add(vec1, *p)` where `p` is a pointer of appropriate vector type. In fact it's not clear if `Load` is actually needed. Only `LoadAligned` is obviously needed because there's no way to encode the fact that the pointer is properly aligned in IL.

fiigii commented:
> Hmm, it does have that - ref. Aka byref in JIT.

But that requires overloading intrincs with `ref` and allow converting `ref element` to/from `ref Vector128/256<T>`.

mikedn commented:
> But that requires overloading intrinsics with ref and allow converting ref element to/from ref Vector128/256<T>.

Only the load/store intrinsics would need that. I think that was discussed but I don't remember how exactly we ended up only with pointer versions. Anyway, doesn't matter to much, they can be added if deemed useful at some point.

fiigii commented:
> Only the load/store intrinsics would need that.

I meant C++ can compile:
```cpp
// vec1 is a __m256 and addr is float*
auto a = _mm256_add_ps(vec1, *(__m256*)addr);
```
to
```asm
vaddps  ymm0, ymm0, ymmword ptr [addr]
```

fiigii commented:
> Only LoadAligned is obviously needed because there's no way to encode the fact that the pointer is properly aligned in IL.

Most of the aligned and unaligned instructions have no difference of performance and function on modern Intel CPUs when the load/store fall into one cache-line. Users should be responsible for alignment of pointers.

mikedn commented:
> I meant C++ can compile:

Yes, and similar code will work in C# except that you can't generate `addps  xmm0, xmmword ptr [addr]` because SSE2 requires memory operands to be aligned and the JIT has no way to know if addr is 16 byte aligned or not.

> But, in C#, we have to rely on folding Load into the comsuer intrinsics.

For SSE2, yes. `Sse2.Add(v1, Sse2.LoadAligned(addr))` should generate `addps xmm0, xmmword ptr[addr]`.

fiigii commented:
>  you can't generate addps xmm0, xmmword ptr [addr] because SSE2 requires memory operands to be aligned and the JIT has no way to know if addr is 16 byte aligned or not.

> Sse2.Add(v1, Sse2.LoadAligned(addr)) should generate addps xmm0, xmmword ptr[addr].

@mikedn `addps` does not require 16-byte alignment. Only **_Type 1_** instructions require explicit alignment:
<img width="856" alt="screen shot 2017-12-17 at 10 56 10 am" src="https://user-images.githubusercontent.com/1263030/34082748-fd483090-e318-11e7-8280-da6ec6ea6d72.png">

tannergooding commented:
@fiigii, there are some instructions, like addsubpd, where they explicitly indicate (on the actual instruction page) that the source operand must be aligned or #GP will be generated.

It would be useful to determine which section of the manual is accurate.

mikedn commented:
Hmm, too lazy to dig through the docs at this hour. Let's test:
```C++
__m128 a[42];
__asm addps xmm0, xmmword ptr [a+1]
```
Crashes with "read access violation". Runs fine with `[a]` or `[a+16]`.
But:
```C++
__m128 a[42];
__asm vaddps xmm0, xmm0, xmmword ptr [a+1]
```
Runs fine.

So yes, classic SSE2 (no VEX encoded) instructions require 16 byte alignment.

fiigii commented:
@mikedn @tannergooding Sorry, I was worry. `Type 2` instructions also require alignment with `legacy SSE encoding`
<img width="835" alt="screen shot 2017-12-17 at 12 08 02 pm" src="https://user-images.githubusercontent.com/1263030/34083307-0898286a-e323-11e7-9957-20b22552540b.png">

fiigii commented:
In summary, we should fold `Load/Store` for all VEX-encoding instructions.
BTW, @mikedn could you point any code that I can learn enabling containment analysis for SIMD operations? I am not very familiar with this part.

mikedn commented:
> could you point any code that I can learn enabling containment analysis for SIMD operations

See the various `ContainCheckX` functions in lowerxarch.cpp. There's `ContainCheckSIMD` but it's rather limited in what it does. See `ContainCheckBinary` too.

In short:
* If the second operand is a memory operand (`GT_IND` usually) you can make it contained by calling `MakeSrcContained`.
* If the second operand is a local variable then you may call `SetRegOptional` to tell the register allocator that it does not need to load the variable into a register (if it's not already in a register).
* If the first operand is a memory operand/local variable and the instruction is commutative then you could try to swap the operands to take advantage of containment
* AFAIR there aren't many, if any, SSE/AVX instruction that have a RMW (read/modify/write) form (e.g. no `addps xmmword ptr [mem], xmm0`). That will simplify things as handling this form is rather convoluted.

tannergooding commented:
Note: CI results are inaccurate until https://github.com/dotnet/coreclr/issues/15618 is resolved as the `IsSupported` check is returning false and no intrinsics actually have code being generated.

I am testing locally to ensure Windows x64 is working, but that is all the validation I am doing at this time.

tannergooding commented:
@fiigii, @mikedn.

Implemented `SSE_Shuffle` using `GT_LIST` for the operands. It's fairly straightforward and can be fairly well isolated from the other code paths.

fiigii commented:
@tannergooding I will change the importer and codgen of HW intrinsic to a table-driven implementation, so that @mikedn's solution ("add a integer data member to GenTreeHWIntrinsic and store the immediate in it.") may be better to unify the exception code.

tannergooding commented:
Implemented `Set`, `SetAll`, and `SetZero` intrinsics.

All that's left now is `Load`, `Store`, and `StaticCast` (and improving codegen in general).

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -103,6 +111,243 @@ void CodeGen::genSSEIntrinsic(GenTreeHWIntrinsic* node)
+
+            argList = argList->Rest();
+            op3     = argList->Current();
+            ival    = op3->AsIntConCommon()->IconValue();
```
This fails for the `Shuffle_r` test, but passes for the `Shuffle_ro` test. Seems to be because `op3` is a `GT_CAST` when optimizations are not enabled.

fiigii replied:
👍

tannergooding replied:
Looks like this is because we aren't forcing expansion of the intrinsics, which causes other downstream processing on the args to occur.

tannergooding replied:
https://github.com/dotnet/coreclr/pull/15639

fiigii commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -405,13 +405,77 @@ GenTree* Compiler::impSSEIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HAND
+            assert(sig->numArgs == 1);
+            op1     = impSIMDPopStack(TYP_FLOAT);
+            retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, intrinsic, TYP_FLOAT, 16);
+            break;
```
I think all the helper intrinsics should be expanded to other "real" intrinsics in the importer. That would simplify the subsequent optimization work.

tannergooding replied:
Do you mean expand to the existing SIMD intrinsics used by `System.Numerics.Vector<T>`?

fiigii replied:
No, I mean helper intrinsic should be converted to other HW intrinsics. For example, `Sse.SetAll(v)` may be expanded to `Sse.Shuffle(aNewLoaclVar(v), 0)`, `Sse.SetZero<float>` may be expanded to `Sse.Xor(v, v)`.

tannergooding commented:
`StaticCast` is implemented. I initially tried to just use `GT_BITCAST`, but ran into a bunch of issues (mostly asserts related to what is/isn't allowed when dealing with TYP_SIMD16). It ended up being significantly easier to just update the lsra for `NI_SSE_StaticCast` to have `isTgtPref = true`.

fiigii commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -103,6 +111,251 @@ void CodeGen::genSSEIntrinsic(GenTreeHWIntrinsic* node)
+            assert(op2 == nullptr);
+            if (op1Reg != targetReg)
+            {
+                emit->emitIns_SIMD_R_R(INS_movaps, targetReg, op1Reg, TYP_SIMD16);
```
I think `NI_SSE_StaticCast` should not reach codgen that should be converted to `CopyBLK` in the importer. Actually, even `copy` may be unnecessary because `Vector128/256<T>` is immutable types.

mikedn replied:
> Actually, even copy may be unnecessary because Vector128/256<T> is immutable types.

Yes, it looks like it's not necessary. It's all `TYP_SIMD16` even if the base type is `float` or `int`.

tannergooding replied:
This is exactly how `GT_BITCAST` works. The codegen is needed for when the `StaticCast` results in assignment rather than immediate use.

For the case of `var y = MyOp(StaticCast(x))`, the JIT will set `op1Reg` and `targetReg` to the same, and consume the value directly (no code will be generated).

However, for the case of `var y = StaticCast(x)`, the JIT has to store the value to the local (assuming it isn't itself optimized away) and must generate code to do the store.

mikedn replied:
> This is exactly how GT_BITCAST works

AFAIK bitcast is intended to allow you to treat a type as another type (e.g. treat `float` as `int`). That is not what `StaticCast` is doing.

tannergooding replied:
See where I set `isTgtPref` in the lsra here: https://github.com/dotnet/coreclr/pull/15538/commits/c75bdb73ef833e96ad6b3b7cc300362fe4518af6#diff-498fd4859cb147c9bf082f5ebf32ca8fR2554

mikedn replied:
I know that but that is not the point. The point is that `StaticCast` should be a no-op at IR level, it's purely a language level construct that allows you to reinterpret `Vector128<float>` as `Vector128<int>`. Due to the way these types are represented in IR it is not the same thing as reinterpreting `float` as `int`.

tannergooding replied:
> AFAIK bitcast is intended to allow you to treat a type as another type (e.g. treat float as int). That is not what StaticCast is doing.

How is that not what `StaticCast` is doing? `StaticCast` is taking a `Vector128<T>` and treating it as a `Vector128<U>`. In both the case of GT_BITCAST and StaticCast, we are taking one set of bits and reinterpreting it.

> should be a no-op

Yes, unless you are storing the value.
```
var x = Sse.SetScalar(4.0f);
var y = Sse.StaticCast<float, int>(x);
x = Sse.SqrtScalar(x);
```

In the above code, we are creating another local and then modifying the first. Therefore, `StaticCast` is not a no-op, it is a store. `y` needs to remain the integer representation of `4.0f` and `x` needs to remain `2.0f`, so that any subsequent code executes correctly.

tannergooding replied:
Ex: The following program
```C++
#include <intrin.h>
#include <iostream>

int main()
{
	auto x = _mm_set_ss(4.0f);
	std::cout << x.m128_f32[0] << ", " << x.m128_f32[1] << ", " << x.m128_f32[2] << ", " << x.m128_f32[3] << "\r\n";

	auto y = _mm_castps_si128(x);
	std::cout << y.m128i_i32[0] << ", " << y.m128i_i32[1] << ", " << y.m128i_i32[2] << ", " << y.m128i_i32[3] << "\r\n";

	x = _mm_sqrt_ss(x);
	std::cout << x.m128_f32[0] << ", " << x.m128_f32[1] << ", " << x.m128_f32[2] << ", " << x.m128_f32[3] << "\r\n";
	std::cout << y.m128i_i32[0] << ", " << y.m128i_i32[1] << ", " << y.m128i_i32[2] << ", " << y.m128i_i32[3] << "\r\n";

    return 0;
}
```

prints:
```
4, 0, 0, 0
1082130432, 0, 0, 0
2, 0, 0, 0
1082130432, 0, 0, 0
```

mikedn replied:
> In both the case of GT_BITCAST and StaticCast, we are taking one set of bits and reinterpreting it.

And reinterpreting basically means doing nothing, the bit pattern does not change. At best, you're copying the bit pattern from a float register to an integer register. And reinterpretation makes sense only when talking about different JIT types - TYP_FLOAT and TYP_INT for example. Otherwise there's nothing to reinterpret.

In the case of SSE the type is the same (it's either TYP_SIMD16 or TYP_SIMD32) and the registers are the same. Keep in mind that `Vector128<float>` and `Vector128<int>` both map to TYP_SIMD16 and Y/XMM registers.

> Yes, unless you are storing the value.

`StaticCast` doesn't store values anywhere. It produces a value that a separate node (a GT_LCL_VAR, GT_IND or another intrinsic node) may store somewhere. Neither `GT_BITCAST` stores anything anywhere.

tannergooding replied:
I'm still not convinced here. SIMD is also doing roughly the same (albeit without setting `isTgtPref` in `lsra`) for the `Vector<T>.op_explicit` calls, which are the equivalent to the `StaticCast` calls here.

I also don't see any existing examples in the importer (although I may have missed them), where the type of reinterpreting you are wanting is being done. Would you be able to point one out?

mikedn replied:
> I'm still not convinced here. SIMD is also doing roughly the same (albeit without setting isTgtPref in lsra)

Leave LSRA alone, focus on the IR. You have a node that produces a TYP_SIMD16 value. You have another node that uses the TYP_SIMD16 value. Why should there be a GT_BITCAST node between those 2 nodes?

> or the Vector<T>.op_explicit calls, which are the equivalent to the StaticCast calls here.

For all I know that code could be useless. It's not the first time when it happens to discover that some piece of JIT code is somehow useless, broken, inefficient etc.

> I also don't see any existing examples in the importer (although I may have missed them), where the type of reinterpreting you are wanting is being done. Would you be able to point one out?

Yeah, I suppose that doesn't exist, it's an unusual case really. I would expect that doing a pop and push of the popped value should work fine.

If it doesn't work then see if you can use a NOP, it's closer to the meaning of `StaticCast` than BITCAST. Besides, I don't think BITCAST has ever been used in HIR so you may encounter problems if you try to use it in the importer.

tannergooding replied:
Just pushing/popping the stack doesn't work, as it detects an unbalanced stack.

Using `gtNewNothingNode` (NOP) doesn't work because the subsequent STLOC instruction expects a GenTree that has the appropriate return type (and gtOp1 needs to carry the value that should be stored).

Returning a NOP node with the appropriate return type fails multiple asserts down the line (such as in `impAssignStructPtr` as it isn't supported and requires more extensive changes.

I am going to keep what I have right now, for now. It is small, simple, produces the correct result, produces (what appears to be) good codegen, and results in a NOP in the places we expect it to. Additionally, similar examples to it can also be found elsewhere in the codebase, so it isn't something new.

mikedn replied:
OK, I'll take a closer look after this is merged. I'm too lazy to get your branch and experiment with it :grin:

tannergooding replied:
To be clear, I'm perfectly happy changing this to something better, it just doesn't appear to be readily supported right now (and I don't want to waste a bunch of time trying to figure it out -- at least not when I was able to follow an existing example and get it working already) 😄

mikedn replied:
> it just doesn't appear to be readily supported right now

Yeah, it could be, especially GT_NOPs with struct types. It's not clear why pop/push doesn't work but that's something that can be investigated in the future.

tannergooding replied:
Push/Pop doesn't work because of the handling in `impImportCall` here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L8199 (callRetType is grabbed from the method sig here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L7047).

We don't have a VOID return type, so we try to push the call node onto the stack here: https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L8199

`impPushOnStack` itself validates that we don't go over `compMaxStack` (https://github.com/dotnet/coreclr/blob/master/src/jit/importer.cpp#L83), which in this case is 1.

fiigii commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -405,13 +405,84 @@ GenTree* Compiler::impSSEIntrinsic(NamedIntrinsic intrinsic, CORINFO_METHOD_HAND
+            assert(sig->numArgs == 1);
+            op1 = impSIMDPopStack(TYP_SIMD16);
+            retNode =
+                gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, intrinsic, getBaseTypeOfSIMDType(sig->retTypeSigClass), 16);
```
Generic HW intrinsics must return `nullptr` if the type arguments are non-numeric types.

tannergooding replied:
Will fix.

tannergooding replied:
Fixed.

tannergooding commented:
> I am so sorry for the delay to review. I am on vacation until Jan 3rd, so future actions may be delayed also ...

@fiigii, don't feel any rush to review. I completely understand that this and the related PRs likely wont get merged until after the new year 😄

mikedn commented:
> don't feel any rush to review. I completely understand that this and the related PRs likely wont get merged until after the new year

Hmm, maybe one of you should merge other's changes into his own fork and then submit a PR containing commits from everyone.

tannergooding commented:
> maybe one of you should merge other's changes into his own fork and then submit a PR containing commits from everyone.

That might work for some of the smaller PRs (which are just making base changes to be consumed by the other PRs), but I don't think that will necessarily work for the larger PRs (like this). It would make them harder to review and harder to get checked in. I think they are all at different stages and are touching different areas (I am working on SSE here, @4creators is doing some work on SSE2 in another PR, etc).

tannergooding commented:
This PR is effectively done.

The remaining intrinsics (`Load`, `LoadAligned`, `Store`, `StoreAligned`, `StoreAlignedNonTemporal`) are going to be implemented separately by @fiigii.

After the remaining intrinsics are done, the instructions implemented here can be improved by updating the containment analysis done for them.

There are a few remaining issues:
1. The hardware intrinsics should be table driven
  A. There is a PR to do that (https://github.com/dotnet/coreclr/pull/15642). Depending on which gets merged first, the other PR will need to be updated accordingly
  B. There is an open PR (https://github.com/dotnet/coreclr/pull/15640) that only contains the first commit of this PR. If this gets merged first, the other can be closed. Otherwise, the first commit should be dropped from this PR
  C. The `Shuffle_r` test is failing. It doesn't actually run in PR due to `COMPlus_EnableIncompleteISAClass` being `0`. There is a PR (https://github.com/dotnet/coreclr/pull/15639) that fixes the issue by forcing intrinsics to always expand (so that we can actually validate the third argument is `GT_CNS_INT`). However, there is still some debate on whether or not the intrinsics should be always expanded or if the instructions that require constant parameters should have a software fallback.

tannergooding commented:
fyi. @dotnet/jit-contrib

tannergooding commented:
Also fyi. @CarolEidt, @RussKeldorph, @eerhardt, since you are the champions for the hw intrinsics feature.

tannergooding commented:
The corlib definitions for the scalar hardware intrinsics got merged in https://github.com/dotnet/coreclr/pull/15341.

Once the CoreFX side gets merged (https://github.com/dotnet/corefx/pull/26095) and gets pumped back, I can add the codegen support for those as well (I have it completed locally, but haven't pushed since it will fail the PR).

tannergooding commented:
There is also the missing `Sse.MoveMask` API that is being added in https://github.com/dotnet/coreclr/pull/15601.

Codegen will need to be added for it after the PR is approved/merged.

tannergooding commented:
FYI. @fiigii 

This now implements all SSE intrinsics (packed and scalar) currently exposed, except for the Load/Store (packed and scalar) that you indicated you would be working on.

Once the Load/Store intrinsics are also implemented, SSE can be marked as fully implemented.

tannergooding commented:
I have a lot of support for containment locally, but I am going to wait until we decide how to extend the HWIntrinsicsInfo table (once everyone gets back from vacation) to get that rebased/pushed.

tannergooding commented:
https://github.com/dotnet/coreclr/pull/15640 was merged, so I dropped the commit I had cherry-picked into this PR.

4creators commented on src/jit/gentree.cpp:
```diff
@@ -6788,6 +6798,16 @@ GenTreeArgList* Compiler::gtNewArgList(GenTreePtr arg1, GenTreePtr arg2, GenTree
 
+/*****************************************************************************
+ *
+ *  Create a list out of the three values.
```
nit: three instead of four

4creators commented on src/jit/emitxarch.cpp:
```diff
@@ -4812,6 +4819,23 @@ void emitter::emitIns_SIMD_R_R_R(instruction ins, regNumber reg, regNumber reg1,
+        {
+            emitIns_R_R(INS_movaps, emitTypeSize(simdtype), reg, reg1);
+        }
+        emitIns_R_R_I(ins, emitTypeSize(simdtype), reg, reg2, ival);
```
Why `reg, reg2` and not `reg, reg1`?

tannergooding replied:
Because `reg` is `reg1` at this point. `ins reg1, reg1, ival` wouldn't do what we need.

4creators commented on src/jit/compiler.h:
```diff
@@ -2088,6 +2107,7 @@ class Compiler
     GenTreeArgList* gtNewArgList(GenTreePtr op);
     GenTreeArgList* gtNewArgList(GenTreePtr op1, GenTreePtr op2);
     GenTreeArgList* gtNewArgList(GenTreePtr op1, GenTreePtr op2, GenTreePtr op3);
+    GenTreeArgList* gtNewArgList(GenTreePtr op1, GenTreePtr op2, GenTreePtr op3, GenTreePtr op4);
```
In the SSE2, AVX and AVX-512 there are 16, 32, and 64 arguments in set_epi8 intrinsic series. What would be the best way to handle them? Creating single GenTreeArgList? Any ideas? I am working now on 16 byte SSE2 set_epi8 and other overloads and it seems there is a need for general design to handle it efficiently.

tannergooding replied:
I think a `GenTreeArgList` is probably the easiest way to handle it today.

Given the number of arguments (and infrequency of intrinics that take more than 4 operands), it is probably best to dynamically create the ArgList using a loop, rather than creating custom constructors for them.

fiigii replied:
We can using something like:
```cpp
GenTree* Compiler::buildArgList(int num, var_types ty)
{
    assert(num >= 2);
    GenTree* list = nullptr;
    for (int i = 0; i < num; i++)
    {
        GenTree* arg = impSIMDPopStack(ty);
        list         = new (this, GT_LIST) GenTreeOp(GT_LIST, ty, arg, list);
    }
    return list;
}
```

4creators replied:
Thanks! How about using an array of GenTreePtr, we recover them anyway while calling impPopStack()?

tannergooding replied:
> How about using an array of GenTreePtr, we recover them anyway while calling 

There are intermediate steps between the creation of the HWIntrinsic node and the codegen phase that would need to be updated to support that, and it may be more trouble than its worth (I had looked at updating some of this so I could have gtOp1 be the first operand and gtOp2 be an ArgList for the remaining, and even that was non-trivial).

> We can have something like:
> `GenTree* Compiler::buildArgList(int num, var_types ty)`

Given that most of the higher argument `Set` intrinsics take a multiple of 4 arguments, we might be able to optimize that a bit more and turn it into a chain of `gtNewArgList(op1, op2, op3, op4)` calls.

fiigii replied:
@4creators I am polishing my table-driven code that already contains this helper function, and I think it can be submitted by the end of tomorrow. Once we have a final solution to the table-driven approach, these infrastructure codes may need to change. I do not want to waste your and @tannergooding's valuable work and time (sorry to block your work), so I recommend to suspend this work for a little while.

4creators replied:
> and it may be more trouble than its worth

If that's the case let's use `GenTreeArgList`.

fiigii replied:
> Given that most of the higher argument Set intrinsics take a multiple of 4 arguments, we might be able to optimize that a bit more and turn it into a chain of gtNewArgList(op1, op2, op3, op4) calls.

`Set` is special, as previously mentioned, these helper-intrinsics can be transformed in the importer, so that I do not need a long arg list actually.

tannergooding replied:
Ah, yes. That's on the list of things I want to fix for this PR 😄

4creators replied:
OK - should I in this case delay implementing `Set` until we have transformation support in the importer or go ahead and later refactor?

4creators replied:
The advantage of implementing now would be that we can get earlier to complete Sse2 ISA implementation.

tannergooding replied:
@4creators, `impSSE2Intrinsic` is part of the importer, so you should just be able to construct and return the appropriate node chain there.

Ex: The following "should" work (currently compiling to validate):
```
case NI_SSE_Set:
{
    assert(sig->numArgs == 4);
    assert(getBaseTypeOfSIMDType(sig->retTypeSigClass) == TYP_FLOAT);

    op4 = impPopStack().val;
    op3 = impPopStack().val;
    op2 = impPopStack().val;
    op1 = impPopStack().val;

    retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, 
        gtNewSimdHWIntrinsicNode(TYP_SIMD16, op4, op3, NI_SSE_UnpackLow, TYP_FLOAT, 16),
        gtNewSimdHWIntrinsicNode(TYP_SIMD16, op2, op1, NI_SSE_UnpackLow, TYP_FLOAT, 16),
        gtNewIconNode(68, TYP_UBYTE),
        NI_SSE_Shuffle, TYP_FLOAT, 16
    );
    break;
}
```

4creators replied:
> impSSE2Intrinsic is part of the importer, so you should just be able to construct and return the appropriate node chain there

OK than it seems that some experiments should not hurt

fiigii replied:
> The advantage of implementing now would be that we can get earlier to complete Sse2 ISA implementation.

Maybe not, we must unify the implementations of all the ISA classes. My plan is
1. complete table-driven infrastructure,
2. finish all the simple intrinsic in AVX/AVX2/SSE/SSE2/SSE41, so I recommend you guys to split the huge PRs.
3. move to complicated guys (e.g., memory access and IMM intrinsics).
4. ...

During this process, if anything blocks your work, we can move to https://github.com/dotnet/coreclr/issues/15490 that looks more independent.

tannergooding commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -412,6 +424,12 @@ void CodeGen::genSSEIntrinsic(GenTreeHWIntrinsic* node)
+        case NI_SSE_SqrtScalar:
+            assert(baseType == TYP_FLOAT);
+            assert(op2 == nullptr);
+            emit->emitIns_SIMD_R_R_R(INS_sqrtss, targetReg, op1Reg, op1Reg, TYP_SIMD16);
```
@fiigii, on the original PR adding the API for these three intrinsics, you recommended having the API be:
`Vector128<float> SqrtScalar(Vector128<float> value)`, rather than 
`Vector128<float> SqrtScalar(Vector128<float> upper, Vector128<float> value)`.

However, this actually means we have to handle these intrinsics specially, since they have a different shape from the others. It also means the user has to inject other intrinsics to simulate the case where `upper` and `value` should not be the same (which most of the other scalar intrinsics already support in some form).

Is there an actual performance benefit to the user doing:
```
var input = Sse.MoveScalar(upper, value);
var result = Sse.SqrtScalar(input);
```

over just providing and allowing:
```
var result = SqrtScalar(upper, value);
```

tannergooding commented:
(thought I posted this already, can't seem to find it, sorry if duplicate)

@fiigii, when initially adding the scalar APIs for `SqrtScalar`, `ReciprocalScalar`, and `ReciprocalSqrtScalar`, you recommended the API be in the form of:
`Vector128<float> SqrtScalar(Vector128<float> value`, rather than `Vector128<float> SqrtScalar(Vector128<float> upper, Vector128<float> value)`.

However, this requires us to handle these intrinsics specially, since they don't have the same shape as the other scalar APIs. Additionally, it means the user has to manually insert additional intrinsics for the case where `upper` and `value` are different.

My question is, is it actually more performant for the user to do:
```
var input = Sse.MoveScalar(upper, value);   // vmovss input, upper, value
var result = Sse.SqrtScalar(input);         // vsqrtss result, input, input
```

than it is to do:
```
var result = Sse.SqrtScalar(upper, value);  // vsqrtss result, upper, value
```

fiigii commented:
> My question is, is it actually more performant for the user to do:

I measured that `vsqrtss result, upper, value` is 3x slower than `vsqrtss result, input, input`, I am not sure with `vmovss`. 
Another problem is this kind of semantics (upper + value) only available with VEX-encoding for most of those instructions, so **if this programming pattern is really needed**, we should expose the `upper + value` version in `Avx` rather than change the `Sse*` APIs.

tannergooding commented:
> I measured that vsqrtss result, upper, value is 3x slower than vsqrtss result, input, input, I am not sure with vmovss.

I think measuring this is worthwhile. I imagine that `vmovss input, upper, value`; `vsrtss result, input, input` performs similarly to `vsqrtss result, upper, value` (as they have to functionally do the same thing), but that the former takes up additional bytes in the pipeline (I don't see anything specific about these in the optimization manual, so I am assuming that the general rule applies).

> Another problem is this kind of semantics (upper + value) only available with VEX-encoding for most of those instructions, so if this programming pattern is really needed, we should expose the upper + value version in Avx rather than change the Sse* APIs.

I think we should be exposing this in both as it will likely lead to better codegen (assuming that `vsqrtss result, upper, value` is better than `vmovss input, upper, value`; `vsqrtss result, input, input`).

In the case of keeping `SqrtScalar(value)`, a user has to code:
```
var input = Sse.MoveScalar(upper, value);
var result = Sse.SqrtScalar(input);
```

This results in the following on VEX:
```
vmovss input, upper, value
vsqrtss result, input, input
```

but the following on non-VEX:
```
movaps input, upper
movss input, value
movaps result, input
sqrtss result, input
```
NOTE: The codegen here could actually match the case below, but would require additional support in the JIT to know that `MoveScalar` followed by `SqrtScalar` should be optimized to it.


However, in the case of providing `SqrtScalar(upper, value)` a user can code:
```
var result = Sse.SqrtScalar(upper, value);
```

This results in the following on VEX:
```
vsqrtss result, upper, value
```

and the following on non-VEX:
```
movaps result, upper
movss result, value
sqrtss result, result
```

fiigii commented:
> I think we should be exposing this in both as it will likely lead to better codegen (assuming that vsqrtss result, upper, value is better than vmovss input, upper, value; vsqrtss result, input, input).

Even though the assumption is supported by some measurement, I believe it is machine(micro-architecture)-dependent.

My suggestion is to leave these APIs to later and to investigate the usefulness of "upper + value" programming pattern in practice.

tannergooding commented:
rebased onto master

tannergooding commented:
I added the code for the `Sse.Shuffle` compiler fallback.

It looks like it isn't quite correct since the test fails with:
```
SSE Shuffle failed on float:
1, -5, 100, 0,
``` 

Going to play around with it some more later this evening to see what's wrong.

4creators commented:
@tannergooding 

I have implemented general Shuffle algorithm which should work with float as well. See: 
https://github.com/dotnetrt/coreclr/blob/c147511372eb8d37f56321798daaa11d1b8c3e16/tests/src/JIT/HardwareIntrinsics/X86/Sse2/Shuffle.cs#L251

`Span<T>` parameters have size of Vector128, another shuffle algo is used for `ShuffleHigh` or `ShuffleLow` - it operates on half of Vector128 length.

tannergooding commented:
> I have implemented general Shuffle algorithm which should work with float as well.

I believe the decision was to not use a software fallback, but rather a compiler fallback (one that still executes the target instruction, via something like a switch table, rather than one that simulates the target instruction via some other implementation).

4creators commented:
> I believe the decision was to not use a software fallback, but rather a compiler fallback 

Yes, you are right. But this code is not a software fallback, it is test logic using in test check method which checks the result of hardware instruction execution by comparing emulation results with hardware instruction results.

tannergooding commented:
For the test logic, we should probably follow the `operation` logic that the Architectures manual gives:
![image](https://user-images.githubusercontent.com/10487869/34701646-465d521a-f49e-11e7-8dd6-813682b96842.png)

4creators commented:
> For the test logic, we should probably follow the operation logic that the Architectures manual gives:

This is exactly the same pseudo code except that it works in loop and in addition checks equivalence between CPU instruction result and operation code result.

tannergooding commented:
Fixed the issue with the `Sse.Shuffle` compiler fallback.

I had missed that the TableDataGen entries were relative to `fgFirstBB` and not relative to the first switch table entry.

tannergooding commented:
Jobs here are failing due to the recent API name changes in https://github.com/dotnet/coreclr/pull/15601.

Will need to wait for https://github.com/dotnet/corefx/pull/26225 merge and pump back into CoreCLR.

tannergooding commented:
Rebased on to current HEAD. Picked up and resolved conflicts from the API name changes and the IsDstDstSrc/IsDstSrcSrc changes

tannergooding commented:
Added the MoveMask intrinsic. Should be back to everything except the Load/Store intrinsics.

tannergooding commented:
Rebased onto master, which brought in the basic containment analysis support.

Latest commit removes code duplication in the codegen where possible by adding a `insOfHWIntrinsic` and `ivalOfHWIntrinsic` method with the same signature as the PR that adds support for driving these via the HWIntrinsicInfo table (https://github.com/dotnet/coreclr/pull/15749).

It also moves all the applicable intrinsics to use the existing containment analysis support for the `R_R_RM` scenario.

tannergooding commented:
FYI. @fiigii, the latest changes should give a better view of the codegen split for the Sse intrinsics.

Going to work on adding containment support for them as well.

tannergooding commented:
Most of the SSE Compare intrinsics support containment now as well.

tannergooding commented:
I ended up doing very basic codegen support for the Load/Store intrinsics. This meant that there is basic codegen support for all declared SSE intrinsics and it could be marked as "fully implemented".

tannergooding commented:
Removed the `Store` intrinsics. There is some additional work I need to figure out there (Checked/Release builds still mark the statements as dead code, so I probably need to mark them with the `GTF_ASG` flag somewhere). The changes so far are in my [sse-intrinsics-store](https://github.com/tannergooding/coreclr/tree/sse-intrinsics-store) branch

CarolEidt commented on src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -188,29 +186,499 @@ void CodeGen::genHWIntrinsic_R_R_RM(GenTreeHWIntrinsic* node, instruction ins)
+            {
+                // We emit a fallback case for the scenario when op3 is not a constant. This should normally
+                // happen when the intrinsic is called indirectly, such as via Reflection. However, it can
+                // also occur if the consumer calls it directly and just doesn't pass a constant value.
```
This looks great, though I'm wondering if it will be productive to try to abstract this so that we don't have to duplicate so much of this for other "immediate-only" instructions.

tannergooding replied:
I think it might be good to abstract it as well.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -23,6 +23,27 @@ extern const char* getHWIntrinsicName(NamedIntrinsic intrinsic)
     return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].intrinsicName;
 }
 
+static const bool isNumericType(var_types type)
```
I wasn't aware that existed 😄

tannergooding replied:
Resolved

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -450,13 +689,238 @@ GenTree* Compiler::impSSEIntrinsic(NamedIntrinsic        intrinsic,
+            op4 = impPopStack().val;
+            op3 = impPopStack().val;
+            op2 = impPopStack().val;
+            op1 = impPopStack().val;
```
Agreed. I held off on the validation here since it looked like @fiigii was going to handle it in his PR adding the table driven support.

tannergooding commented:
@fiigii, @CarolEidt. How do we want to do this to minimize conflict with the table driven PR?

Do we want to merge this now and resolve conflicts in the other PR, or do we want to hold off on merging this until the latter is merged.

CarolEidt commented:
> How do we want to do this to minimize conflict with the table driven PR? Do we want to merge this now and resolve conflicts in the other PR, or do we want to hold off on merging this until the latter is merged.

I think that is best determined by you and @fiigii

fiigii commented:
I am okay to merge this first and solve conflicts in the table-driven PR.

tannergooding commented:
Had to resolve a merge conflict with the System.Math.Round intrinsic PR (https://github.com/dotnet/coreclr/pull/14736).

Went ahead and added the function header comments and changed `isNumericType` to be `varTypeIsArithmetic` instead.

Will merge once tests are green.

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -152,6 +173,224 @@ InstructionSet Compiler::isaOfHWIntrinsic(NamedIntrinsic intrinsic)
     return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].isa;
 }
 
+int Compiler::ivalOfHWIntrinsic(NamedIntrinsic intrinsic)
```
Resolved

sandreenko commented:
Do not want to bother you, but frankejit is broken again:
`f:\codegenmirror\src\ndp\clr\src\jit\emitxarch.cpp(4140): error C3861: 'emitHandleMemOp': identifier not found [F:\CodegenMirror\src\NDP\clr\src\jit\frankenjit\frankenjit.nativeproj]`

So new functions should be under `#ifndef LEGACY_BACKEND`

tannergooding commented:
> Do not want to bother you, but frankejit is broken again:

No worries. I should have remembered I needed to `ifdef` the R_R_A_I method after just fixing the others.

Fix should be https://github.com/dotnet/coreclr/pull/15900. I tested locally on the TFS sources and it looks like JIT builds succesfully.
